remote: add AddToStoreNar operation

squalus opened the pull request:
Some open items:

- What type should be used for the reply? Nothing is sent over the wire. `()` is already taken by `opSuccess` which reads a boolean. Maybe `opSuccess` should use a new reply type, and this one should use `()`
- The client chooses the chunk size. What should that be?
- Better errors for the `RemoteStoreError_Fixme` usages
- Make an Arbitrary instance for `SampleNar`

Closes #265

sorki commented:
Very nice (especially the framing!) and sorry for the conflicts! Wanted to merge earlier but didn't get to it sooner.

The `CheckSignatures` is also in core (I've realized it is actually `Check`Signatures as well only couple days ago) - but don't worry about it, I'll deal with it.

You're right about `opSuccess`, most ops simply return `True` and indicate errors via logger so it seemed appropriate to just use `()` but this one makes it weird, I agree about the newtype.

No idea about the chunk size tbh.

sorki commented on hnix-store-remote/src/System/Nix/Store/Remote/Client/Core.hs:
```diff
@@ -75,14 +81,57 @@ doReq = \case
+            if narBytes == 0 then writeFramedSource (const (pure Nothing)) soc 0
+            else throwError RemoteStoreError_NoDataSourceProvided
+          Just dataSource -> do
+            writeFramedSource dataSource soc narBytes
```
Now I kinda wonder what's the `Logger_Read` code path for.

squalus replied:
Yeah, I don't know if it's still used in the newer protocol versions.

squalus commented:
- Replaced the existing `()` reply type with a new `SuccessCodeReply` type
- Added a `NoReply` type for this case
- Added error types
- Resolved conflicts

sorki commented:
Nicely done, thank you!

sorki commented:
Btw I'm pretty much done with most large-scale changes/refactors so things should settle down now, feel free to tackle what you see fit.
